src/mgr/QueryParams.cc

```diff
@@ -68,60 +69,61 @@ Mgr::QueryParams::find(const String& name) const
     return iter;
 }
 
-bool
-Mgr::QueryParams::ParseParam(SBuf &paramStr, Param &param)
+/// Parses the value part of a "param=value" URL section.
+/// Value can be a comma separated list of integers, or an opaque string.
+Mgr::QueryParam::Pointer
+ParseParamValue(SBuf &valueStr)
 {
-    static const CharacterSet nameChars = CharacterSet("param-name", "_") + CharacterSet::ALPHA + CharacterSet::DIGIT;
-
-    Parser::Tokenizer tok(paramStr);
-
-    SBuf name;
-    if (!tok.prefix(name, nameChars))
-        return false;
-
-    if (!tok.skip('='))
-        return false; // TODO: support name-only parameters
-
-    static const CharacterSet badValueChars = CharacterSet("param-value-bad", "&= ");
-    SBuf valueStr = tok.remaining();
-    if (valueStr.findFirstOf(badValueChars) != SBuf::npos)
-        return false;
-
-    param.first = SBufToString(name);
+    static const CharacterSet comma("comma",",");
 
+    Parser::Tokenizer tok(valueStr);
     std::vector<int> array;
     int64_t intVal = 0;
     while (tok.int64(intVal, 10, false)) {
-        static const CharacterSet comma("comma",",");
         (void)tok.skipAll(comma);
         Must(intVal >= std::numeric_limits<int>::min());
         Must(intVal <= std::numeric_limits<int>::max());
-        array.emplace_back(int(intVal));
+        array.emplace_back(intVal);
     }
 
     if (tok.atEnd())
-        param.second = new IntParam(array);
+        return new Mgr::IntParam(array);
     else
-        param.second = new StringParam(SBufToString(valueStr));
-
-    return true;
+        return new Mgr::StringParam(SBufToString(valueStr));
 }
 
-bool
+/**
+ * Syntax:
+ *   query  = param *( '&' param ) [ '#' fragment ]
+ *   param  = name '=' value
+ *   name   = [a-zA-Z0-9]+
+ *   value  = *pchar | ( 1*DIGIT *( ',' 1*DIGIT ) )
+ */
+void
 Mgr::QueryParams::Parse(Parser::Tokenizer &tok, QueryParams &aParams)
 {
-    static const CharacterSet paramDelim("query-param","&");
-
-    SBuf foundStr;
-    while (tok.token(foundStr, paramDelim)) {
-        Param param;
-        if (!ParseParam(foundStr, param))
-            return false;
-        aParams.params.push_back(param);
-    }
+    static const CharacterSet nameChars = CharacterSet("param-name", "_") + CharacterSet::ALPHA + CharacterSet::DIGIT;
+    static const CharacterSet valueChars = CharacterSet("param-value", "&= #").complement();
+
+    while (!tok.atEnd() && !tok.skip('#')) {
 
-    // success == no garbage in query-string
-    return tok.atEnd();
+        SBuf nameStr;
+        if (!tok.prefix(nameStr, nameChars))
+            throw TexcHere("invalid character in action name");
+        if (!tok.skip('='))
+            throw TexcHere("missing parameter value");
+
+        SBuf valueStr;
+        if (!tok.prefix(valueStr, valueChars))
+            throw TexcHere("invalid character in parameter value");
+
+        if (tok.skip('=') || tok.skip(' '))
+            throw TexcHere("invalid character in parameter value");
+
```

Comment:
Unnecessary leftovers?

```suggestion
```

Reply:
No. These two characters being seen are an error. If we ignore them existing here, they would be treated as a new name by the next loop iteration.

Reply:
> No. These two characters being seen are an error.

Accepting my suggestion would not change that fact.

> If we ignore them existing here, they would be treated as a new name by the next loop iteration.

And that is exactly what should happen! They will be seen as a start of a new (invalid) parameter name. There is nothing special about these two invalid characters -- many characters cannot be a part of a parameter name (or value). To have a better chance at being correct, a hand-written parser must focus on correctly identifying all valid input. Double checking random invalid cases distracts from that focus and, sooner or later, leads to bugs.

At this point, the code has already parsed the parameter value. Do not second-guess that code. Move on to the next construct to be parsed.

Reply:
The "next construct to be parsed" is one of the delimiter characters: 
 * '&' - next param
 * '#' - start fragment
 * '=' - bad baseN encoding terminator
 * SP - bad URI %-encoding
 * NUL - parse complete (aka atEnd())

Reply:
The "bad" items on the list do not exist in the grammar. In theory, one could defend their presence on the grounds of making error messages better or providing other human-oriented enhancements, but I urge you not to change the excuse to that one. With so many bugs, we should be focusing on the essentials instead of poorly arguing about often buggy bells and whistles (which will, in turn, need polishing, causing more arguments, etc.). Writing parsers by code review is painful enough. Let's not make it worse.

The good items on the list are indeed a part of the grammar, but they are a red herring for this change request. They are (or will be, once the code is fixed) handled. This change request is not asking to remove their handling. This change request is specific to an explicit test for two bad items that are not a part of the grammar.

Reply:
Done in 1d7af98